Sources/SwiftGenKit/Parsers/Strings/StringsDictEntry.swift

```diff
@@ -0,0 +1,212 @@
+    let valueTypeKey: String
+    let zero: String?
+    let one: String?
+    let two: String?
+    let few: String?
+    let many: String?
+    let other: String
+
+    private enum CodingKeys: String, CodingKey {
+      case specTypeKey = "NSStringFormatSpecTypeKey"
+      case valueTypeKey = "NSStringFormatValueTypeKey"
+
+      case zero, one, two, few, many, other
+    }
+  }
+}
+
+// - MARK: Decodable conformance
+
+extension StringsDict: Decodable {
+  // Custom CodingKeys struct to be able to parse dictionaries with dynamic keys.
+  // This is necessary, because each variable used in the formatKey is used as the key
+  // to another dictionary containing the rules for that variable.
+  private struct CodingKeys: CodingKey {
+    var intValue: Int?
+    var stringValue: String
+
+    init?(intValue: Int) {
+      self.intValue = intValue
+      self.stringValue = "\(intValue)"
+    }
+    init?(stringValue: String) {
+      self.stringValue = stringValue
+    }
+    init(key: String) {
+      self.stringValue = key
+    }
+
+    // fixed keys
+    static let formatKey = CodingKeys(key: "NSStringLocalizedFormatKey")
+    static let variableWidthRules = CodingKeys(key: "NSStringVariableWidthRuleType")
+  }
+
+  init(from coder: Decoder) throws {
+    let container = try coder.container(keyedBy: CodingKeys.self)
+    let formatKey = try container.decodeIfPresent(String.self, forKey: .formatKey)
+    let variableWidthRules = try container.decodeIfPresent([String: String].self, forKey: .variableWidthRules)
+
+    // We either have a formatKey OR we have a variableWidthRule.
+    switch (formatKey, variableWidthRules) {
+    case (.none, .none), (.some, .some):
+      throw Strings.ParserError.invalidFormat
+    case let (.some(formatKey), .none):
+      let variables = try StringsDict.decodeVariableRules(in: container, formatKey: formatKey)
+      self = .pluralEntry(PluralEntry(formatKey: formatKey, variables: variables))
+    case let (.none, .some(variableWidthRules)):
+      self = .variableWidthEntry(VariableWidthEntry(rules: variableWidthRules))
+    }
+  }
+
+  private static func decodeVariableRules(
+    in container: KeyedDecodingContainer<StringsDict.CodingKeys>,
+    formatKey: String
+  ) throws -> [PluralEntry.Variable] {
+    guard let variableKeyTuples = StringsDict.variableKeysFromFormatKey(formatKey) else { return [] }
+    let sortedVariableKeys = variableKeyTuples
+      .sorted { lhs, rhs in
+        // Sort by positional argument if present, otherwise by occurrence in the format key
+        switch (lhs.positionalArgument, rhs.positionalArgument, lhs.range.lowerBound, rhs.range.lowerBound) {
+        case let (.none, .none, lhs, rhs):
+          return lhs < rhs
+        case (.some, .none, _, _):
+          return true
+        case (.none, .some, _, _):
+          return false
+        case let (.some(lhs), .some(rhs), _, _):
+          return lhs < rhs
```

Comment:
This seems to sort by occurence in the formatKey first, and only then by positionalArgument, while the comment says the opposite. Not sure which should be the right one.

Also the case `.some, .none, _, _` means that a positional argument is always before a non-positional one, but in `Name %@ value %2$d` the non-positional should be before the positional one 🤔

Reply:
No, it always prefers positional arguments over non-positional ones, since the `case let (.none, .none, lhs, rhs):` is only evaluated if both arguments of the comparison are non-positional ones.

Yes, correct. We probably shouldn't even support a mix of positional and non-positional arguments. I just tried out to generate something like your example and this would only work (as in not-crash) if the passed CVarArgs are in the exact same order as the placeholders. E.g.
- `Name %@ value %2$d` ✅
- `Name %@ value %3$d` ✅ (It just takes the arguments by passed order if the positional numbers don't fit)
- `Name %@ value %@ out of %2$d`  ❌
- `Name %@ value %1$d`  ❌

I'd propose to treat a mix of positional and non-positional arguments as programmer-error and log a warning in this case.

Reply:
I agree that we should treat invalid cases (like your last 2 examples) as programmer error and log. But a mix of positional and non-positional like your first example should be valid and accepted (it already is accepted by our current code when used in `.strings` files so no reason we should reject it if used in `.stringsdict` files)

I remember struggling when I had to implement `placeholders(fromFormat: )` to handle a lot of edge cases like invalid formats or formats with either "holes" in certain positions or referenced multiple times (sometimes with positional info explicit sometimes inferred) having to match types (for example it should be valid to have `Name %@ value %.2f more precise value %2$.6f` I think because here the 2nd parameter is expected to be a floating point value in both implicit and explicit position 2)… not an easy problem to solve I agree.

Reply:
The invalid cases throw errors and the valid cases are working and generate the correct code now.

I also added tests for interleaved positional and non-positional placeholders, as they are supported by Apple. The only condition is that the number of the lowest positional placeholder needs to be larger than the count of non-positional placeholders.